### PR TITLE
Quadlet container mount - support non key=val options

### PR DESCRIPTION
**Pull request by ygalblum**
Some keys, e.g. ro do not have values.
The current implementation crashed looking for the = sign

Externalize findMountType in a new package
Parse mount command using FindMountType    
Add test case and adjust the test framework

#### Does this PR introduce a user-facing change?
Yes

```release-note
Quadlet container mount - support non key=val options
```

Resolves: #20104

### CODE REVIEW

**Comment by ygalblum** [comment]
Sorry for opening another PR for this issue while #20108 and #20112 are open. I wasn't sure if I can push to these branches.
This change addresses the example issue in #20104 and as raised by @Luap99 [here](https://github.com/containers/podman/pull/20112#issuecomment-1733234399)

**Comment by ygalblum** [comment]
@vrothberg @rhatdan @Luap99 @guillaumerose PTAL

**Review by vrothberg — COMMENTED** [review]
I think that the underlying problem of having two separate functions to parse essentially the same thing remains.  That implies a risk of code divergence and subtle errors.

Did you consider recycling the Podman-internal function to parse mounts and change it in a way that we can re-use it in Quadlet?

**Comment by ygalblum** [comment]
> I think that the underlying problem of having two separate functions to parse essentially the same thing remains. That implies a risk of code divergence and subtle errors.
> 
> Did you consider recycling the Podman-internal function to parse mounts and change it in a way that we can re-use it in Quadlet?

My idea was to limit the parsing to only what Quadlet really needs and that is resolving the `.volume` case. Unlike the original implementation all the rest is not looked or changed. As such the assumptions are only that:
1. The parameter separator it `,`
2. `type`, `source` and `src` are in the `key=value` format

I think that reusing the Podman parser code might raise other issues. For example, Quadlet is not supposed to do any validity checks. But, now, it will.

**Comment by Luap99** [comment]
> I think that the underlying problem of having two separate functions to parse essentially the same thing remains. That implies a risk of code divergence and subtle errors.
> 
> Did you consider recycling the Podman-internal function to parse mounts and change it in a way that we can re-use it in Quadlet?

Exactly, --mount flag uses csv not split at comma, this behaves differently for special chars so it is important to use the same code. And for the future someone might change the --mount parsing and then will not update quadlet so it will diverge again. The only way to address this is to directly share the code.

**Comment by Luap99** [comment]
you can just use https://github.com/containers/podman/blob/a2434a9d374da3e77bc8d94e684f610c94242bf2/pkg/specgenutil/volumes.go#L145

That should already work for both use cases podman and quadlet. Simply move this function to a new package without all the extra imports as that would bloat quadlet otherwise.

**Comment by ygalblum** [comment]
@vrothberg @Luap99 Done. PTAL

File: pkg/systemd/quadlet/quadlet.go
Comment (Luap99) [inline pkg/systemd/quadlet/quadlet.go:1551]: This will still fail when the value contains special chars, you must again encode this via the csv lib in order to encode this back into the right format. Consider a a filepath with comma:
`--mount=type=bind,src=/tmp,"dst=/path,1"`

**Review by vrothberg — APPROVED** [review]
LGTM

File: pkg/systemd/quadlet/quadlet.go
Comment (vrothberg) [reply pkg/systemd/quadlet/quadlet.go:1551]: Great catch.  Once fixed, can we also add a test case for it?

Comment (ygalblum) [reply pkg/systemd/quadlet/quadlet.go:1551]: Adding the test case is the first step in writing the fix :)

Comment (vrothberg) [reply pkg/systemd/quadlet/quadlet.go:1551]: Like Batman, you are not the maintainer we deserve but the one we need :^)

Comment (ygalblum) [reply pkg/systemd/quadlet/quadlet.go]: Fixed. Wow, making the test work was more complicated than the code fix :)

**Comment by ygalblum** [comment]
Does this failure block the merge: `bin/quadlet grew by 53951 bytes; max allowed is 51200.`?

**Comment by Luap99** [comment]
> Does this failure block the merge: `bin/quadlet grew by 53951 bytes; max allowed is 51200.`?

Yes but we can overwrite with the label, given this is so close I am good with it. You already split out the function to not require unnecessary imports so I assume this is the best we can do and the grow is mostly caused by the CSV parser.

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Review by vrothberg — APPROVED** [review]
LGTM, thanks @ygalblum !
I let @Luap99 have a final head nod

**Comment by Luap99** [comment]
/lgtm

Thanks

**Pull request merged**
